wellcomeml/ml/clustering.py

```diff
@@ -0,0 +1,321 @@
+
+        # Linearises Dictionary to be compatible with grid search so it
+        # becomes one dictionary with 'step__parameter'
+        vectorizer_grid = ParameterGrid(
+            param_grid=param_grid.get('vectorizer', {})
+        )
+
+        reducer_grid = ParameterGrid(
+            param_grid=param_grid.get('reducer', {})
+        )
+
+        clustering_grid = ParameterGrid(
+            param_grid=param_grid.get('clustering', {})
+        )
+
+        best_silhouette = -1
+        params_list = []
+        best_params = None
+
+        logging_level = logger.level
+        if verbose <= 1:
+            # Previously disable logging to allow the loading bar to run
+            # uninterruptly. Will reset after.
+            logging.getLogger().setLevel(logging.WARNING)
+            logger.setLevel(logging.WARNING)
+
+        for vectorizer_param in vectorizer_grid:
+            self.vectorizer.set_params(**vectorizer_param)
+            self._fit_step(X, step='vectorizer')
+            for reducer_param in reducer_grid:
+                self.reducer.set_params(**reducer_param)
+                self._fit_step(step='reducer')
+                for clustering_param in clustering_grid:
+                    self.clustering.set_params(**clustering_param)
+                    self._fit_step(step='clustering')
+
+                    silhouette = _clustering_score(self.clustering,
+                                                   self.reduced_points)
+                    n_clusters = len(np.unique(self.clustering.labels_))
+                    noise = sum(self.clustering.labels_ == -1)
+
+                    # If there is noise, the label "-1" is added as a
+                    # valid cluster number, so we need to subtract it
+
+                    n_clusters -= 1*(noise > 0)
+                    params = {
+                        'vectorizer': vectorizer_param,
+                        'reducer': reducer_param,
+                        'clustering': clustering_param
+                    }
+                    params_list += [{
+                        "params": params,
+                        "silhouette": silhouette,
+                        "n_clusters": n_clusters,
+                        "noise": noise
+                    }]
+
+                    if silhouette > best_silhouette and min_n_clusters <= \
+                            n_clusters <= max_n_clusters and noise < max_noise:
+                        best_silhouette = silhouette
+                        best_params = deepcopy(params)
+
+        # Fits the pipeline again with the best parameters
+        self.set_params(best_params)
+        self.fit(X)
+
+        logger.setLevel(logging_level)
+
+        return {
+            "params_list": params_list,
+            "silhouette": best_silhouette,
+            "best_params": best_params
+        }
+
+    def stability(self):
+        """Function to calculate how stable the clusters are"""
```

Comment:
FYI @nonieA , will look into implementing stability in the future. Happy to have your contribution here if you want.

Reply:
that will be a nice addition